diff --git a/src/shared/dbus-client.h b/src/shared/dbus-client.h
index 6d2dd49..dbe4531 100644
--- a/src/shared/dbus-client.h
+++ b/src/shared/dbus-client.h
@@ -12,6 +12,7 @@
 #ifndef BLUEALSA_SHARED_DBUSCLIENT_H_
 #define BLUEALSA_SHARED_DBUSCLIENT_H_
 
+#include <endian.h>
 #include <poll.h>
 #include <stddef.h>
 #include <stdint.h>
@@ -170,12 +171,23 @@ struct ba_pcm {
 
 	/* 16-bit packed PCM volume */
 	union {
+#if __BYTE_ORDER == __LITTLE_ENDIAN
 		struct {
 			dbus_uint16_t ch2_volume:7;
 			dbus_uint16_t ch2_muted:1;
 			dbus_uint16_t ch1_volume:7;
 			dbus_uint16_t ch1_muted:1;
 		};
+#elif __BYTE_ORDER == __BIG_ENDIAN
+		struct {
+			dbus_uint16_t ch1_muted:1;
+			dbus_uint16_t ch1_volume:7;
+			dbus_uint16_t ch2_muted:1;
+			dbus_uint16_t ch2_volume:7;
+		};
+#else
+#error "Unknown byte order"
+#endif
 		dbus_uint16_t raw;
 	} volume;
